### PR TITLE
regex-utils: Add support for translating regex character sets into wildcards when possible.

### PR DESCRIPTION
**Pull request by Bill-hbrhbr**
# Description
Implements the following functionality:
  * Reduces a character set into a single character if possible.
    * A trivial character set containing a single character or a single escaped metacharacter.
      * E.g. `[a]` into `a`, `[\^]` into `^`
    * If the `case_insensitive_wildcard` config is turned on, the translator can also reduce the
      case-insensitive style character set patterns into a single lowercase character:
      * E.g. `[aA]` into `a`, `[Bb]` into `b`, `[xX][Yy][zZ]` into `xyz`

# Validation performed
Verified in unit tests

### CODE REVIEW

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]
Before going to the detailed review, I had some comments about the way of tracking the charset. In the current implementation, we have an iterator in the `TranslatorState` and expose APIs to update/query the underlying iterator that points to the beginning of the charset. However, it's a little confusing and unsafe:
1. The constructor also takes an iterator, which doesn't make sense if it is not the beginning of the charset.
2. The names of `m_it`, `get_marked_iterator`, and `mark_iterator` don't really indicate their usage.
3. When querying an iterator, we don't have a sanity check of whether it is set to the beginning of a charset properly.

I would propose a solution which we can add a dedicated class to track the status of charset escape. It should have a counter of chars processed (indicating the length of the charset), a flag indicating whether it's an escape char or not, and also the storage of the character to proceed. The `TranslatorState` should have a `std::optional` member of such a charset status tracker, and emplace the optional tracker when `[` is found and reset the optional tracker when `]` is found. All the checks can be passed down to the charset tracker's member function to proceed when you see the character in `charset_state_transition`, and you don't need the chaining `if-else` with the length check. Defininig sub-states inside this charset tracker class might be useful. I think that would make the interface more clean and readable.

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 1]
```diff
@@ -183,14 +247,74 @@ auto escaped_state_transition(
...
+    char parsed_char{};
+
+    if (1 == charset_len) {
+        if (cCharsetNegate == ch0 || cEscapeChar == ch0) {
+            return ErrorCode::UnsupportedCharsetPattern;
+        }
+        parsed_char = ch0;
+    } else {  // 2 == charset_len
+        if (cEscapeChar == ch0 && cRegexCharsetEscapeSeqMetaCharsLut.at(ch1)) {
+            // 2-char escape sequence
```
Comment (davidlion) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:290]:
```suggestion
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 2]
```diff
@@ -103,6 +114,25 @@ using StateTransitionFuncSig
...
 
+/**
+ * Attempts to reduce regex character sets into a single character so that the regex string is still
+ * translatable to wildcard.
+ *
+ * In most cases, only a trival character set containing a single character is reducable. However,
+ * if the output wildcard query will be analyzed in case-insensitive mode, character set patterns
+ * such as [aA] [Bb] are also reducable.
+ * Throws two possible kinds of error codes, with IncompleteCharsetStructure having a higher
+ * precedence over UnsupportedCharsetPattern.
```
Comment (davidlion) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:125]:
```suggestion
* On error returns either IncompleteCharsetStructure or UnsupportedCharsetPattern.
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 3]
```diff
@@ -183,14 +247,74 @@ auto escaped_state_transition(
...
+        }
+        parsed_char = ch0;
+    } else {  // 2 == charset_len
+        if (cEscapeChar == ch0 && cRegexCharsetEscapeSeqMetaCharsLut.at(ch1)) {
+            // 2-char escape sequence
+            parsed_char = ch1;
+        } else if (config.case_insensitive_wildcard()
+                   && matching_upper_lower_case_char_pair(ch0, ch1))
+        {
+            // case-insensitive patterns like [aA] [Bb] etc.
```
Comment (davidlion) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:295]:
```suggestion
```

**Review by davidlion — CHANGES_REQUESTED** [review]

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 4]
```diff
@@ -48,12 +57,23 @@ class TranslatorState {
...
+    auto set_charset_start(string_view::const_iterator charset_start) -> void {
+        m_charset_start = charset_start;
+    }
+
+    auto invalidate_charset_start() -> void { m_charset_start = nullopt; }
+
 private:
     // Members
     RegexPatternState m_state{RegexPatternState::Normal};
+    optional<string_view::const_iterator> m_charset_start{nullopt};
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:76]:
```suggestion
    optional<string_view::const_iterator> m_charset_start;
```
You can rely on the default constructor

Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:76]: How about `m_charset_start_it`?

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 5]
```diff
@@ -48,12 +57,23 @@ class TranslatorState {
...
+    }
+
     // Setters
     auto set_next_state(RegexPatternState const& state) -> void { m_state = state; }
 
+    auto set_charset_start(string_view::const_iterator charset_start) -> void {
+        m_charset_start = charset_start;
+    }
+
+    auto invalidate_charset_start() -> void { m_charset_start = nullopt; }
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:71]:
```suggestion
    auto invalidate_charset_start() -> void { m_charset_start.reset(); }
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 6]
```diff
@@ -183,14 +255,74 @@ auto escaped_state_transition(
...
+        string_view::const_iterator& it,
+        string& wildcard_str,
+        RegexToWildcardTranslatorConfig const& config
+) -> error_code {
+    auto const charset_start_opt{state.get_charset_start()};
+    if (false == charset_start_opt.has_value()) {
+        return ErrorCode::IllegalState;
     }
-    wildcard_str += ch;
+    string_view::const_iterator const charset_start = charset_start_opt.value();
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:273]:
```suggestion
    auto const charset_start = charset_start_opt.value();
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 7]
```diff
@@ -215,6 +347,10 @@ auto final_state_cleanup(
             // multichar wildcard
             wildcard_str += cSingleCharWildcard;
             break;
+        case TranslatorState::RegexPatternState::Charset:
+        case TranslatorState::RegexPatternState::CharsetEscaped:
+            return ErrorCode::IncompleteCharsetStructure;
+            break;
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:353]: Do we need `break` after return?

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 8]
```diff
@@ -114,6 +152,36 @@ using StateTransitionFuncSig
  */
 [[nodiscard]] StateTransitionFuncSig final_state_cleanup;
 
+// Other helpers
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:155]: Not necessary
```suggestion
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 9]
```diff
@@ -114,6 +152,36 @@ using StateTransitionFuncSig
...
+    }
+    wildcard_str += ch;
+}
+
+/**
+ * Detects if the two input arguments are a matching pair of upper and lowercase characters.
+ *
+ * @param ch0
+ * @param ch1
+ * @return True if the input is a matching pair.
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:176]:
```suggestion
 * @return Whether the input chars are a matching pair.
```

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 10]
```diff
@@ -114,6 +152,36 @@ using StateTransitionFuncSig
...
+// Other helpers
+/**
+ * Appends a single character as a literal to the wildcard string.
+ *
+ * If the literal is a metacharacter in the wildcard syntax, prepend the literal with an escape
+ * backslash.
+ * @param ch The literal to be appended.
+ * @param wildcard_str The wildcard string to be updated.
+ */
+inline auto append_char_to_wildcard(char const ch, string& wildcard_str) -> void {
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:164]:
```suggestion
auto append_char_to_wildcard(char const ch, string& wildcard_str) -> void {
```
1. We don't explicitly add `inline` unless the profiling shows a different on performance
2. We should put helper declarations at the top and move the actual implementation to the bottom, inside the anon namespace.

File: components/core/src/clp/regex_utils/regex_translation_utils.cpp [hunk 11]
```diff
@@ -114,6 +152,36 @@ using StateTransitionFuncSig
...
+}
+
+/**
+ * Detects if the two input arguments are a matching pair of upper and lowercase characters.
+ *
+ * @param ch0
+ * @param ch1
+ * @return True if the input is a matching pair.
+ */
+inline auto matching_upper_lower_case_char_pair(char const ch0, char const ch1) -> bool {
```
Comment (LinZhihao-723) [inline components/core/src/clp/regex_utils/regex_translation_utils.cpp:178]:
```suggestion
[[nodiscard]] auto matching_upper_lower_case_char_pair(char const ch0, char const ch1) -> bool {
```

**Review by LinZhihao-723 — CHANGES_REQUESTED** [review]

**Review by davidlion — APPROVED** [review]

**Review by LinZhihao-723 — APPROVED** [review]

**Pull request merged**